<!DOCTYPE html>
<html lang="ko">
   <head>
      <meta charset="UTF-8" />
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <title>Document</title>
   </head>
   <body>
      <script>
         function asyncTask() {
            return new Promise((resolve, reject) => {
               setTimeout(() => {
                  console.log('Task done')
                  resolve('Task result')
               }, 1000)
            })
         }

         // async: 비동기 함수로 만든다
         async function runAsyncTasks() {
            // try안에서 실행하는 코드에서 에러가 발생하면 catch로 에러 메세지가 전달된다.
            try {
               // await: async 함수 내부에서만 사용 가능, promise의 then과 같은 역할(순서를 만들어준다)
               // await를 사용시 asyncTask() 함수가 다 끝날때까지 기다렸다가 그 다음 라인의 코드를 실행한다
               const result1 = await asyncTask()
               console.log(result1)

               const result2 = result1 + 'next'
               console.log(result2)
            } catch (error) {
               console.error(error)
            }
         }

         runAsyncTasks() // 함수 실행
         console.log('Task out')
      </script>
   </body>
</html>
